File: GhoulCipher.py
from functools import reduce
import base64 as b64


def b64e(s):
    return b64.b64encode(s.encode()).decode()


def b64d(s):
    return b64.b64decode(s).decode()


def get_xor_key(key):
    return reduce(lambda x, y: x ^ y, key.encode())


def xor_cipher(message, key):
    return bytes([m ^ key[i % len(key)] for i, m in enumerate(message)])


def alg(m, xor_key, p, k):
    temp = p - k ** p * p ** k + p - p ** k
    return bytearray(
        (i ^ (temp % (k + 1)) ^ xor_key + (p % k)) % k
        for i in m
    )


def ghoul_cipher(message, p, k, key):
    message = bytearray(
        (m + p ** k + k ** p - p) % k
        for m in xor_cipher(message.encode(), key.encode())
    )
    return alg(message, get_xor_key(key), p, k)


def ghoul_decipher(cipher_message, p, k, key):
    cipher_message = bytearray(
        (cm - p ** k - k ** p + p) % k
        for cm in alg(cipher_message, get_xor_key(key), p, k)
    )
    try:
        cipher_message = xor_cipher(cipher_message, key.encode()).decode()
    except UnicodeDecodeError:
        return None
    return cipher_message


"""
def main():
  message = ".❤️.......................937, Sin que alguien dijera nada, sin dar su nombre, como un rompecabezas resolviéndose por sí mismo. De4th==■ Con una mirada fría y vivida, estaba el Dios de la muerte. 𝔤𝔥𝔬𝔲𝔩..."
  k = 104729
  n = 256
  # LLave generada aleatoriamente
  key = "frU0v^5MG%b@uHhQ5n5!jQ%Ok^BlVaFM4#^6HEDU"

  print(message)
  cm = ghoul_cipher(b64e(message), k, n, key)
  #print(list(cm))
  print(''.join(list(chr(c % n) for c in list(cm))))
  dm = ghoul_decipher(cm, k, n, "frU0v^5MG%b@uHhQ5n5!jQ%Ok^BlVaFM4#^6HEDU")
  print(b64d(dm) if dm is not None else "Clave Incorrecta")

if __name__ == '__main__':
  main()
"""

File: static/dashboard/js/ghoul-cipher.js
const b64e = (s) => {
  return btoa(s).toString();
};

const b64d = (s) => {
  return atob(s).toString();
};

function get_xor_key(key) {
  let xorKey = 0;
  for (let i = 0; i < key.length; i++) {
    xorKey ^= key.charCodeAt(i);
  }
  return xorKey;
}

function xor_cipher(message, key) {
  return message.map((byte, index) => byte ^ key.charCodeAt(index % key.length));
}

function alg(m, xorKey, p, k) {
  let temp = p - Math.pow(k, p) * Math.pow(p, k) + p - Math.pow(p, k);
  let result = [];
  for (let i = 0; i < m.length; i++) {
    result.push((m[i] ^ temp % (k + 1) ^ (xorKey + (p % k))) % k);
  }
  return new Uint8Array(result);
}

function ghoul_cipher(message, p, k, key) {
  let encodedMessage = [];
  let xorKey = get_xor_key(key);
  for (let i = 0; i < message.length; i++) {
    let cipheredByte = (message[i] + Math.pow(p, k) + Math.pow(k, p) - p) % k;
    encodedMessage.push(cipheredByte);
  }
  let cipheredMessage = xor_cipher(encodedMessage, key);
  return cipheredMessage;
}

function ghoul_decipher(cipher_message, p, k, key) {
  let xorKey = get_xor_key(key);
  let decodedMessage = xor_cipher(cipher_message, key);
  let decipheredMessage = [];
  for (let i = 0; i < decodedMessage.length; i++) {
    decipheredMessage.push((decodedMessage[i] - xorKey - (p % k) + k) % k);
  }
  try {
    decipheredMessage = new Uint8Array(decipheredMessage);
    return decipheredMessage.reduce(
      (str, byte) => str + String.fromCharCode(byte),
      ""
    );
  } catch (error) {
    return null;
  }
}
